### PR TITLE
[SPARK-31845][CORE][TESTS] Refactor DAGSchedulerSuite by introducing completeAndCheckAnswer and using completeNextStageWithFetchFailure

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
**First**
`DAGSchedulerSuite` provides `completeNextStageWithFetchFailure` to make all tasks in non first stage occur `FetchFailed`.
But many test case uses complete directly as follows:
```scala
 complete(taskSets(1), Seq(
     (FetchFailed(makeBlockManagerId("hostA"),
        shuffleDep1.shuffleId, 0L, 0, 0, "ignored"), null)))
```
We need to reuse `completeNextStageWithFetchFailure`.

**Second**
`DAGSchedulerSuite` also check the results show below:
```scala
complete(taskSets(0), Seq((Success, 42)))
assert(results === Map(0 -> 42))
```
We can extract it as a generic method of `checkAnswer`.

### Why are the changes needed?
Reuse `completeNextStageWithFetchFailure`

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 1]
```diff
@@ -405,6 +405,15 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     runEvent(JobCancelled(jobId, None))
   }
 
+  /** Make task set success and check result. */
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:408]: It doesn't make the whole taskset success but only some tasks in the taskset success?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -405,6 +405,15 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+  private def checkAnswer(
+    taskSet: TaskSet,
+    results: Seq[(TaskEndReason, Any)],
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:411]: I think the name of `results` is not appropriate here considering its usage and also confused with `this.results` below. Maybe, `taskEndInfos`?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 3]
```diff
@@ -405,6 +405,15 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
     runEvent(JobCancelled(jobId, None))
   }
 
+  /** Make task set success and check result. */
+  private def checkAnswer(
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:409]: Maybe, `completeAndCheckAnswer`?

**Review by Ngone51 — COMMENTED** [review]
Seems like `checkAnswer` only has one line change for the developer. But overall, I think it's good to have it.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 4]
```diff
@@ -405,6 +405,15 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+    taskSet: TaskSet,
+    results: Seq[(TaskEndReason, Any)],
+    expected: Map[Int, Any]): Unit = {
```
[the remark this answers is not in the document]
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:412]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 1 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:408]: You said right, I will change the comments.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 3 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:409]: OK

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 2 above)
Comment (beliefer) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala:411]: Let's update `results` to `taskEndInfos` for `complete` too.

**Comment by beliefer** [comment]
retest this please

**Comment by Ngone51** [comment]
retest this please

**Comment by Ngone51** [comment]
LGTM, also cc @jiangxb1987

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @beliefer and @Ngone51 .
Merged to master.

**Comment by beliefer** [comment]
@dongjoon-hyun @Ngone51 Thanks for your help!

**Pull request closed without merging**